// @istvankreisz/copdeck-scraper-app/dist/browserConfigs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.browsers = void 0;
var chrome = {
    name: 'Chrome',
    headers: {
        stockx1: {
            accept: 'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9',
            'accept-language': 'en-US,en;q=0.9',
            'cache-control': 'no-cache',
            pragma: 'no-cache',
            'sec-ch-ua': '" Not;A Brand";v="99", "Google Chrome";v="91", "Chromium";v="91"',
            'sec-ch-ua-mobile': '?0',
            'sec-fetch-dest': 'document',
            'sec-fetch-mode': 'navigate',
            'sec-fetch-site': 'none',
            'sec-fetch-user': '?1',
            'upgrade-insecure-requests': '1',
            referrerPolicy: 'strict-origin-when-cross-origin',
            mode: 'cors',
        },
        stockx2: {
            accept: '*/*',
            'accept-language': 'en-US,en;q=0.9',
            appos: 'web',
            appversion: '0.1',
            authorization: '',
            'cache-control': 'no-cache',
            pragma: 'no-cache',
            'sec-ch-ua': '" Not;A Brand";v="99", "Google Chrome";v="91", "Chromium";v="91"',
            'sec-ch-ua-mobile': '?0',
            'sec-fetch-dest': 'empty',
            'sec-fetch-mode': 'cors',
            'sec-fetch-site': 'same-origin',
            'x-requested-with': 'XMLHttpRequest',
            referrer: 'https://stockx.com/search',
            referrerPolicy: 'same-origin',
            mode: 'cors',
        },
        stockx3: {
            accept: '*/*',
            'accept-language': 'en-US,en;q=0.9',
            appos: 'web',
            appversion: '0.1',
            authorization: '',
            'cache-control': 'no-cache',
            pragma: 'no-cache',
            'sec-ch-ua': '" Not;A Brand";v="99", "Google Chrome";v="91", "Chromium";v="91"',
            'sec-ch-ua-mobile': '?0',
            'sec-fetch-dest': 'empty',
            'sec-fetch-mode': 'cors',
            'sec-fetch-site': 'same-origin',
            'x-requested-with': 'XMLHttpRequest',
            referrer: 'https://stockx.com/',
            referrerPolicy: 'same-origin',
            mode: 'cors',
        },
        goat1: {
            accept: 'application/json',
            'accept-language': 'en-US,en;q=0.9',
            'cache-control': 'no-cache',
            'content-type': 'application/x-www-form-urlencoded',
            pragma: 'no-cache',
            'sec-ch-ua': '" Not;A Brand";v="99", "Google Chrome";v="91", "Chromium";v="91"',
            'sec-ch-ua-mobile': '?0',
            'sec-fetch-dest': 'empty',
            'sec-fetch-mode': 'cors',
            'sec-fetch-site': 'cross-site',
            referrer: 'https://www.goat.com/',
            referrerPolicy: 'strict-origin-when-cross-origin',
            mode: 'cors',
        },
        goat2: {
            accept: 'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9',
            'accept-language': 'en-US,en;q=0.9',
            'cache-control': 'no-cache',
            pragma: 'no-cache',
            'sec-ch-ua': '" Not;A Brand";v="99", "Google Chrome";v="91", "Chromium";v="91"',
            'sec-ch-ua-mobile': '?0',
            'sec-fetch-dest': 'document',
            'sec-fetch-mode': 'navigate',
            'sec-fetch-site': 'none',
            'sec-fetch-user': '?1',
            'upgrade-insecure-requests': '1',
            referrerPolicy: 'strict-origin-when-cross-origin',
            mode: 'cors',
        },
        goat3: {
            accept: 'application/json',
            'accept-language': 'en-US,en;q=0.9',
            'cache-control': 'no-cache',
            pragma: 'no-cache',
            'sec-ch-ua': '" Not;A Brand";v="99", "Google Chrome";v="91", "Chromium";v="91"',
            'sec-ch-ua-mobile': '?0',
            'sec-fetch-dest': 'empty',
            'sec-fetch-mode': 'cors',
            'sec-fetch-site': 'same-origin',
            referrer: 'https://www.goat.com/sneakers/',
            referrerPolicy: 'strict-origin-when-cross-origin',
            mode: 'cors',
        },
        klekt1: {
            accept: 'application/json, text/plain, */*',
            'accept-language': 'en-US,en;q=0.9',
            'cache-control': 'no-cache',
            'content-type': 'application/json;charset=UTF-8',
            pragma: 'no-cache',
            'sec-ch-ua': '" Not;A Brand";v="99", "Google Chrome";v="91", "Chromium";v="91"',
            'sec-ch-ua-mobile': '?0',
            'sec-fetch-dest': 'empty',
            'sec-fetch-mode': 'cors',
            'sec-fetch-site': 'same-site',
            referrer: 'https://www.klekt.com/',
            referrerPolicy: 'strict-origin-when-cross-origin',
            mode: 'cors',
        },
        klekt2: {
            accept: 'application/json, text/plain, */*',
            'accept-language': 'en-US,en;q=0.9',
            'cache-control': 'no-cache',
            'content-type': 'application/json;charset=UTF-8',
            pragma: 'no-cache',
            'sec-ch-ua': '" Not;A Brand";v="99", "Google Chrome";v="91", "Chromium";v="91"',
            'sec-ch-ua-mobile': '?0',
            'sec-fetch-dest': 'empty',
            'sec-fetch-mode': 'cors',
            'sec-fetch-site': 'same-site',
            referrer: 'https://www.klekt.com/',
            referrerPolicy: 'strict-origin-when-cross-origin',
            mode: 'cors',
        },
    },
    userAgent: 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.77 Safari/537.36',
};
var firefox = {
    name: 'Firefox',
    headers: {
        stockx1: {
            credentials: 'omit',
            Accept: 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8',
            'Accept-Language': 'en-US,en;q=0.5',
            'Upgrade-Insecure-Requests': '1',
            Pragma: 'no-cache',
            'Cache-Control': 'no-cache',
            mode: 'cors',
        },
        stockx2: {
            Host: 'stockx.com',
            Accept: '*/*',
            'Accept-Language': 'en-US,en;q=0.5',
            'Accept-Encoding': 'gzip, deflate, br',
            Referer: 'https://stockx.com/search',
            'X-Requested-With': 'XMLHttpRequest',
            Authorization: '',
            appVersion: '0.1',
            appOS: 'web',
            DNT: '1',
            Connection: 'keep-alive',
            Pragma: 'no-cache',
            'Cache-Control': 'no-cache',
        },
        stockx3: {
            credentials: 'include',
            Accept: '*/*',
            'Accept-Language': 'en-US,en;q=0.5',
            'X-Requested-With': 'XMLHttpRequest',
            Authorization: '',
            appVersion: '0.1',
            appOS: 'web',
            Pragma: 'no-cache',
            'Cache-Control': 'no-cache',
            referrer: 'https://stockx.com/',
            mode: 'cors',
        },
        goat1: {
            Host: '2fwotdvm2o-dsn.algolia.net',
            Accept: 'application/json',
            'Accept-Language': 'en-US,en;q=0.5',
            'Accept-Encoding': 'gzip, deflate, br',
            'Content-Type': 'application/x-www-form-urlencoded',
            Origin: 'https://www.goat.com',
            Referer: 'https://www.goat.com/',
            Connection: 'keep-alive',
            Pragma: 'no-cache',
            'Cache-Control': 'no-cache',
        },
        goat2: {
            Host: 'www.goat.com',
            Accept: 'application/json',
            'Accept-Language': 'en-US,en;q=0.5',
            'Accept-Encoding': 'gzip, deflate, br',
            DNT: '1',
            Connection: 'keep-alive',
            'Upgrade-Insecure-Requests': '1',
            Pragma: 'no-cache',
            'Cache-Control': 'no-cache',
        },
        goat3: {
            Host: 'www.goat.com',
            Accept: 'application/json',
            'Accept-Language': 'en-US,en;q=0.5',
            'Accept-Encoding': 'gzip, deflate, br',
            DNT: '1',
            Connection: 'keep-alive',
            Referer: 'https://www.goat.com/sneakers/',
            Pragma: 'no-cache',
            'Cache-Control': 'no-cache',
        },
        klekt1: {
            Host: 'apiv2.klekt.com',
            Accept: 'application/json, text/plain, */*',
            'Accept-Language': 'en-US,en;q=0.5',
            'Accept-Encoding': 'gzip, deflate, br',
            'Content-Type': 'application/json;charset=utf-8',
            Origin: 'https://www.klekt.com',
            DNT: '1',
            Connection: 'keep-alive',
            Referer: 'https://www.klekt.com/',
            Pragma: 'no-cache',
            'Cache-Control': 'no-cache',
        },
        klekt2: {
            Host: 'apiv2.klekt.com',
            Accept: 'application/json, text/plain, */*',
            'Accept-Language': 'en-US,en;q=0.5',
            'Accept-Encoding': 'gzip, deflate, br',
            'Content-Type': 'application/json;charset=utf-8',
            Origin: 'https://www.klekt.com',
            DNT: '1',
            Connection: 'keep-alive',
            Referer: 'https://www.klekt.com/',
            Pragma: 'no-cache',
            'Cache-Control': 'no-cache',
        },
    },
    userAgent: 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:89.0) Gecko/20100101 Firefox/89.0',
};
var opera = {
    name: 'OPR',
    headers: {
        stockx1: {
            accept: 'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9',
            'accept-language': 'en-US,en;q=0.9',
            'sec-fetch-dest': 'document',
            'sec-fetch-mode': 'navigate',
            'sec-fetch-site': 'none',
            'sec-fetch-user': '?1',
            'upgrade-insecure-requests': '1',
            referrerPolicy: 'no-referrer-when-downgrade',
            mode: 'cors',
        },
        stockx2: {
            accept: '*/*',
            'accept-language': 'en-US,en;q=0.9',
            appos: 'web',
            appversion: '0.1',
            authorization: '',
            'sec-fetch-dest': 'empty',
            'sec-fetch-mode': 'cors',
            'sec-fetch-site': 'same-origin',
            'x-requested-with': 'XMLHttpRequest',
            referrer: 'https://stockx.com/search',
            referrerPolicy: 'same-origin',
            mode: 'cors',
        },
        stockx3: {
            accept: '*/*',
            'accept-language': 'en-US,en;q=0.9',
            appos: 'web',
            appversion: '0.1',
            authorization: '',
            'sec-fetch-dest': 'empty',
            'sec-fetch-mode': 'cors',
            'sec-fetch-site': 'same-origin',
            'x-requested-with': 'XMLHttpRequest',
            referrer: 'https://stockx.com/',
            referrerPolicy: 'same-origin',
            mode: 'cors',
        },
        goat1: {
            Host: '2fwotdvm2o-dsn.algolia.net',
            Connection: 'keep-alive',
            accept: 'application/json',
            'content-type': 'application/x-www-form-urlencoded',
            Origin: 'https://www.goat.com',
            'Sec-Fetch-Site': 'cross-site',
            'Sec-Fetch-Mode': 'cors',
            'Sec-Fetch-Dest': 'empty',
            Referer: 'https://www.goat.com/search?query=nike',
            'Accept-Encoding': 'gzip, deflate, br',
            'Accept-Language': 'en-US,en;q=0.9',
        },
        goat2: {
            'upgrade-insecure-requests': '1',
            referrerPolicy: 'no-referrer-when-downgrade',
            mode: 'cors',
        },
        goat3: {
            accept: 'application/json',
            'accept-language': 'en-US,en;q=0.9',
            'sec-fetch-dest': 'empty',
            'sec-fetch-mode': 'cors',
            'sec-fetch-site': 'same-origin',
            referrer: 'https://www.goat.com/sneakers/',
            referrerPolicy: 'no-referrer-when-downgrade',
            mode: 'cors',
        },
        klekt1: {
            accept: 'application/json, text/plain, */*',
            'accept-language': 'en-US,en;q=0.9',
            'content-type': 'application/json;charset=UTF-8',
            'sec-fetch-dest': 'empty',
            'sec-fetch-mode': 'cors',
            'sec-fetch-site': 'same-site',
            referrer: 'https://www.klekt.com/',
            referrerPolicy: 'no-referrer-when-downgrade',
            mode: 'cors',
        },
        klekt2: {
            accept: '*/*',
            'accept-language': 'en-US,en;q=0.9',
            'sec-fetch-dest': 'empty',
            'sec-fetch-mode': 'cors',
            'sec-fetch-site': 'same-site',
            referrer: 'https://www.klekt.com/us/product/',
            referrerPolicy: 'no-referrer-when-downgrade',
            mode: 'cors',
        },
    },
    userAgent: 'Mozilla/5.0 (Macintosh; Intel Mac OS X 11_0_0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/87.0.4280.88 Safari/537.36 OPR/73.0.3856.329',
};
var safari = {
    name: 'Safari',
    headers: {
        stockx1: {
            Accept: 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
            Pragma: 'no-cache',
            'Cache-Control': 'no-cache',
            Host: 'stockx.com',
            'Accept-Language': 'en-us',
            'Accept-Encoding': 'gzip, deflate, br',
            Connection: 'keep-alive',
        },
        stockx2: {
            Pragma: 'no-cache',
            Accept: '*/*',
            Authorization: '',
            'Accept-Language': 'en-us',
            'Accept-Encoding': 'gzip, deflate, br',
            'Cache-Control': 'no-cache',
            Host: 'stockx.com',
            Referer: 'https://stockx.com/search',
            Connection: 'keep-alive',
            'X-Requested-With': 'XMLHttpRequest',
            appVersion: '0.1',
            appOS: 'web',
        },
        stockx3: {
            Pragma: 'no-cache',
            Accept: '*/*',
            Authorization: '',
            'Accept-Language': 'en-us',
            'Accept-Encoding': 'gzip, deflate, br',
            'Cache-Control': 'no-cache',
            Host: 'stockx.com',
            Referer: 'https://stockx.com/',
            Connection: 'keep-alive',
            'X-Requested-With': 'XMLHttpRequest',
            appVersion: '0.1',
            appOS: 'web',
        },
        goat1: {
            'Content-Type': 'application/x-www-form-urlencoded',
            Pragma: 'no-cache',
            Accept: 'application/json',
            'Accept-Language': 'en-us',
            'Accept-Encoding': 'gzip, deflate, br',
            'Cache-Control': 'no-cache',
            Host: '2fwotdvm2o-dsn.algolia.net',
            Origin: 'https://www.goat.com',
            Connection: 'keep-alive',
            Referer: 'https://www.goat.com/',
        },
        goat2: {
            Accept: 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
            Pragma: 'no-cache',
            'Cache-Control': 'no-cache',
            Host: 'www.goat.com',
            'Accept-Language': 'en-us',
            'Accept-Encoding': 'gzip, deflate, br',
            Connection: 'keep-alive',
        },
        goat3: {
            Accept: 'application/json',
            Pragma: 'no-cache',
            'Cache-Control': 'no-cache',
            'Accept-Language': 'en-us',
            Host: 'www.goat.com',
            Referer: 'https://www.goat.com/sneakers',
            'Accept-Encoding': 'gzip, deflate, br',
            Connection: 'keep-alive',
        },
        klekt1: {
            'Content-Type': 'application/json;charset=utf-8',
            Pragma: 'no-cache',
            Accept: 'application/json, text/plain, */*',
            'Accept-Language': 'en-us',
            'Accept-Encoding': 'gzip, deflate, br',
            'Cache-Control': 'no-cache',
            Host: 'apiv2.klekt.com',
            Origin: 'https://www.klekt.com',
            Connection: 'keep-alive',
            Referer: 'https://www.klekt.com/category/all',
        },
        klekt2: {
            'Content-Type': 'application/json;charset=utf-8',
            Pragma: 'no-cache',
            Accept: 'application/json, text/plain, */*',
            'Accept-Language': 'en-us',
            'Accept-Encoding': 'gzip, deflate, br',
            'Cache-Control': 'no-cache',
            Host: 'apiv2.klekt.com',
            Origin: 'https://www.klekt.com',
            Connection: 'keep-alive',
            Referer: 'https://www.klekt.com',
        },
    },
    userAgent: 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0.2 Safari/605.1.15',
};
var browsers = [safari, chrome, opera, firefox];
exports.browsers = browsers;
