// @istvankreisz/copdeck-scraper-app/dist/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExchangeRates = exports.Proxy = exports.StorePrices = exports.InventoryItem = exports.PriceAlertSchema = exports.PriceAlert = exports.ItemSchema = exports.Item = exports.StoreInfoSchema = exports.StoreInfo = exports.ALLCURRENCIESINTERNAL = exports.CurrencyCodeInternal = exports.CurrencyInternal = exports.ALLCURRENCIES = exports.NOK = exports.CHF = exports.GBP = exports.USD = exports.EUR = exports.CurrencyCode = exports.Currency = exports.ALLSTORES = exports.GOAT = exports.KLEKT = exports.STOCKX = exports.Store = exports.ALLFEETYPES = exports.FeeType = exports.SELL = exports.BUY = exports.ALLCOUNTRIES = exports.ConditionEnum = exports.Condition = exports.ItemStatus = exports.Optional = void 0;
var superstruct_1 = require("superstruct");
var Optional = function (value) {
    return superstruct_1.nullable(superstruct_1.optional(value));
};
exports.Optional = Optional;
/////////////////////////
// Country
/////////////////////////
var Austria = { code: 'AT', name: 'Austria' };
var Belgium = { code: 'BE', name: 'Belgium' };
var Bulgaria = { code: 'BG', name: 'Bulgaria' };
var Croatia = { code: 'HR', name: 'Croatia' };
var Cyprus = {
    code: 'CY',
    name: 'Republic of Cyprus',
};
var CzechRepublic = {
    code: 'CZ',
    name: 'Czech Republic',
};
var Denmark = { code: 'DK', name: 'Denmark' };
var Estonia = { code: 'EE', name: 'Estonia' };
var Finland = { code: 'FI', name: 'Finland' };
var France = { code: 'FR', name: 'France' };
var Germany = { code: 'DE', name: 'Germany' };
var Greece = { code: 'GR', name: 'Greece' };
var Hungary = { code: 'HU', name: 'Hungary' };
var Iceland = { code: 'IS', name: 'Iceland' };
var Ireland = { code: 'IE', name: 'Ireland' };
var Italy = { code: 'IT', name: 'Italy' };
var Latvia = { code: 'LV', name: 'Latvia' };
var Liechtenstein = { code: 'LI', name: 'Liechtenstein' };
var Lithuania = { code: 'LT', name: 'Lithuania' };
var Luxembourg = { code: 'LU', name: 'Luxembourg' };
var Malta = { code: 'MT', name: 'Malta' };
var Netherlands = { code: 'NL', name: 'Netherlands' };
var Norway = { code: 'NO', name: 'Norway' };
var Poland = { code: 'PL', name: 'Poland' };
var Portugal = { code: 'PT', name: 'Portugal' };
var Romania = { code: 'RO', name: 'Romania' };
var Slovakia = { code: 'SK', name: 'Slovakia' };
var Slovenia = { code: 'SI', name: 'Slovenia' };
var Spain = { code: 'ES', name: 'Spain' };
var Sweden = { code: 'SE', name: 'Sweden' };
var Switzerland = { code: 'CH', name: 'Switzerland' };
var UK = { code: 'GB', name: 'UK' };
var US = { code: 'US', name: 'US (mainland)' };
var USAH = {
    code: 'US-AH',
    name: 'US (Alaska, Hawaii)',
};
var ALLCOUNTRIES = [
    Austria,
    Belgium,
    Bulgaria,
    Croatia,
    Cyprus,
    CzechRepublic,
    Denmark,
    Estonia,
    Finland,
    France,
    Germany,
    Greece,
    Hungary,
    Iceland,
    Ireland,
    Italy,
    Latvia,
    Liechtenstein,
    Lithuania,
    Luxembourg,
    Malta,
    Netherlands,
    Norway,
    Poland,
    Portugal,
    Romania,
    Slovakia,
    Slovenia,
    Spain,
    Sweden,
    Switzerland,
    UK,
    US,
    USAH,
];
exports.ALLCOUNTRIES = ALLCOUNTRIES;
/////////////////////////
/////////////////////////
/////////////////////////
// FeeType
/////////////////////////
var FeeType = superstruct_1.union([superstruct_1.literal('None'), superstruct_1.literal('Buy'), superstruct_1.literal('Sell')]);
exports.FeeType = FeeType;
var NONE = 'None';
var BUY = 'Buy';
exports.BUY = BUY;
var SELL = 'Sell';
exports.SELL = SELL;
var ALLFEETYPES = [NONE, BUY, SELL];
exports.ALLFEETYPES = ALLFEETYPES;
/////////////////////////
/////////////////////////
/////////////////////////
// Store
/////////////////////////
var STOCKX = { id: 'stockx', name: 'StockX' };
exports.STOCKX = STOCKX;
var KLEKT = { id: 'klekt', name: 'Klekt' };
exports.KLEKT = KLEKT;
var GOAT = { id: 'goat', name: 'GOAT' };
exports.GOAT = GOAT;
var ALLSTORES = [STOCKX, KLEKT, GOAT];
exports.ALLSTORES = ALLSTORES;
var Store = superstruct_1.union([
    superstruct_1.object({ id: superstruct_1.literal(STOCKX.id), name: superstruct_1.literal(STOCKX.name) }),
    superstruct_1.object({ id: superstruct_1.literal(KLEKT.id), name: superstruct_1.literal(KLEKT.name) }),
    superstruct_1.object({ id: superstruct_1.literal(GOAT.id), name: superstruct_1.literal(GOAT.name) }),
]);
exports.Store = Store;
/////////////////////////
// Currency
/////////////////////////
var EUR = { code: 'EUR', symbol: '€' };
exports.EUR = EUR;
var USD = { code: 'USD', symbol: '$' };
exports.USD = USD;
var GBP = { code: 'GBP', symbol: '£' };
exports.GBP = GBP;
var CHF = { code: 'CHF', symbol: 'CHF' };
exports.CHF = CHF;
var NOK = { code: 'NOK', symbol: 'NOK' };
exports.NOK = NOK;
/////// Regular ////////
var ALLCURRENCIES = [EUR, USD, GBP];
exports.ALLCURRENCIES = ALLCURRENCIES;
var Currency = superstruct_1.union([
    superstruct_1.object({ code: superstruct_1.literal(EUR.code), symbol: superstruct_1.literal(EUR.symbol) }),
    superstruct_1.object({ code: superstruct_1.literal(USD.code), symbol: superstruct_1.literal(USD.symbol) }),
    superstruct_1.object({ code: superstruct_1.literal(GBP.code), symbol: superstruct_1.literal(GBP.symbol) }),
]);
exports.Currency = Currency;
var CurrencyCode = superstruct_1.union([superstruct_1.literal(EUR.code), superstruct_1.literal(USD.code), superstruct_1.literal(GBP.code)]);
exports.CurrencyCode = CurrencyCode;
/////// Internal ////////
var ALLCURRENCIESINTERNAL = [EUR, USD, GBP, CHF, NOK];
exports.ALLCURRENCIESINTERNAL = ALLCURRENCIESINTERNAL;
var CurrencyInternal = superstruct_1.union([
    superstruct_1.object({ code: superstruct_1.literal(EUR.code), symbol: superstruct_1.literal(EUR.symbol) }),
    superstruct_1.object({ code: superstruct_1.literal(USD.code), symbol: superstruct_1.literal(USD.symbol) }),
    superstruct_1.object({ code: superstruct_1.literal(GBP.code), symbol: superstruct_1.literal(GBP.symbol) }),
    superstruct_1.object({ code: superstruct_1.literal(CHF.code), symbol: superstruct_1.literal(CHF.symbol) }),
    superstruct_1.object({ code: superstruct_1.literal(NOK.code), symbol: superstruct_1.literal(NOK.symbol) }),
]);
exports.CurrencyInternal = CurrencyInternal;
var CurrencyCodeInternal = superstruct_1.union([
    superstruct_1.literal(EUR.code),
    superstruct_1.literal(USD.code),
    superstruct_1.literal(GBP.code),
    superstruct_1.literal(CHF.code),
    superstruct_1.literal(NOK.code),
]);
exports.CurrencyCodeInternal = CurrencyCodeInternal;
/////////////////////////
// ItemStatus
/////////////////////////
var ItemStatus = superstruct_1.union([
    superstruct_1.object({
        listingPrices: superstruct_1.defaulted(superstruct_1.array(superstruct_1.object({
            store: Store,
            price: superstruct_1.number(),
        })), function () { return []; }),
    }),
    superstruct_1.object({
        sellingPrice: Optional(superstruct_1.number()),
        store: Optional(Store),
    }),
]);
exports.ItemStatus = ItemStatus;
/////////////////////////
/////////////////////////
/////////////////////////
// Condition
/////////////////////////
var Condition = superstruct_1.enums(['new', 'used']);
exports.Condition = Condition;
var ConditionEnum;
(function (ConditionEnum) {
    ConditionEnum["new"] = "new";
    ConditionEnum["used"] = "used";
})(ConditionEnum || (ConditionEnum = {}));
exports.ConditionEnum = ConditionEnum;
/////////////////////////
/////////////////////////
/////////////////////////
// StoreInfo
/////////////////////////
var StoreInfoSchema = {
    name: superstruct_1.string(),
    sku: superstruct_1.string(),
    slug: superstruct_1.string(),
    retailPrice: Optional(superstruct_1.number()),
    imageURL: Optional(superstruct_1.string()),
    url: superstruct_1.string(),
    sellUrl: superstruct_1.string(),
    buyUrl: superstruct_1.string(),
    productId: Optional(superstruct_1.string()),
    brand: superstruct_1.string(),
    store: Store,
};
exports.StoreInfoSchema = StoreInfoSchema;
var StoreInfo = superstruct_1.object(StoreInfoSchema);
exports.StoreInfo = StoreInfo;
/////////////////////////
/////////////////////////
/////////////////////////
// InventoryItem
/////////////////////////
var InventoryItem = superstruct_1.object({
    size: superstruct_1.string(),
    currencyCode: CurrencyCode,
    lowestAsk: Optional(superstruct_1.object({
        noFees: superstruct_1.number(),
        withSellerFees: Optional(superstruct_1.number()),
        withBuyerFees: Optional(superstruct_1.number()),
    })),
    highestBid: Optional(superstruct_1.object({
        noFees: superstruct_1.number(),
        withSellerFees: Optional(superstruct_1.number()),
        withBuyerFees: Optional(superstruct_1.number()),
    })),
    shoeCondition: Optional(superstruct_1.string()),
    boxCondition: Optional(superstruct_1.string()),
    tags: superstruct_1.defaulted(superstruct_1.array(superstruct_1.string()), function () { return []; }),
});
exports.InventoryItem = InventoryItem;
/////////////////////////
/////////////////////////
/////////////////////////
// Store Prices
/////////////////////////
var StorePrices = superstruct_1.object({
    store: Store,
    retailPrice: Optional(superstruct_1.number()),
    inventory: superstruct_1.array(InventoryItem),
});
exports.StorePrices = StorePrices;
/////////////////////////
/////////////////////////
/////////////////////////
// Item
/////////////////////////
var ItemSchema = {
    id: superstruct_1.string(),
    name: Optional(superstruct_1.string()),
    retailPrice: Optional(superstruct_1.number()),
    imageURL: Optional(superstruct_1.object({
        url: superstruct_1.string(),
        store: Store,
    })),
    // name: string(),
    // brand: optional(string()),
    // retailPrice: optional(number()),
    // releaseDate: optional(date()),
    ownedByCount: Optional(superstruct_1.number()),
    priceAlertCount: Optional(superstruct_1.number()),
    storeInfo: superstruct_1.array(StoreInfo),
    storePrices: superstruct_1.array(StorePrices),
    created: Optional(superstruct_1.number()),
    updated: Optional(superstruct_1.number()),
};
exports.ItemSchema = ItemSchema;
var Item = superstruct_1.object(ItemSchema);
exports.Item = Item;
/////////////////////////
/////////////////////////
/////////////////////////
// Price Alert
/////////////////////////
var PriceAlertSchema = {
    id: superstruct_1.string(),
    itemId: superstruct_1.string(),
    name: superstruct_1.string(),
    targetPrice: superstruct_1.number(),
    targetSize: superstruct_1.string(),
    relation: superstruct_1.union([superstruct_1.literal('above'), superstruct_1.literal('below')]),
    feeType: FeeType,
    priceType: superstruct_1.union([superstruct_1.literal('ask'), superstruct_1.literal('bid')]),
    stores: superstruct_1.array(Store),
    lastNotificationSent: Optional(superstruct_1.number()),
    created: Optional(superstruct_1.number()),
    updated: Optional(superstruct_1.number()),
};
exports.PriceAlertSchema = PriceAlertSchema;
var PriceAlert = superstruct_1.object(PriceAlertSchema);
exports.PriceAlert = PriceAlert;
/////////////////////////
/////////////////////////
/////////////////////////
// Proxy
/////////////////////////
var Proxy = superstruct_1.object({
    host: superstruct_1.string(),
    port: superstruct_1.number(),
    protocol: superstruct_1.string(),
    auth: superstruct_1.optional(superstruct_1.object({
        username: superstruct_1.string(),
        password: superstruct_1.string(),
    })),
});
exports.Proxy = Proxy;
/////////////////////////
/////////////////////////
/////////////////////////
// ExchangeRates
/////////////////////////
var ExchangeRates = superstruct_1.object({
    usd: superstruct_1.number(),
    gbp: superstruct_1.number(),
    chf: superstruct_1.number(),
    nok: superstruct_1.number(),
});
exports.ExchangeRates = ExchangeRates;
